### ISSUE TITLE
Access Denied - This event has been logged

### ISSUE THREAD

**Issue by DrMMu** [ampache/ampache#2277]
## Description

Sometimes I can't access my Ampache 4.1.0 server (I had the same problem with 4.0.4, but never had it with 3.9.0).
I'm using DSub on Android to play music, and it will say that the server is unreachable. If I try to log via any browser, it says:

>Access Denied
>This event has been logged
>
>You have been redirected to this page because you do not have access to this function
>
>If you believe this is an error please contact an Ampache administrator
>
>This event has been logged

See this screenshot (for Ampache 4.0.4, for 4.1.0 it looks identical):
![Ampache 4 0 4-release Login Error](https://user-images.githubusercontent.com/33975632/71766734-4b1fbd00-2f03-11ea-8109-971fb39cd04b.png)

I don't know how to reproduce this, I'm afraid. :-(
I'm basically using two different users, one admin user and one "user" user, and I've got somehow the feeling that it has to do with me logging in simultaneously from different machines with the same user, but this is a wild guess at best.

The error is identical to #2025 resp. #2027, but according to the discussion there this (1) should have been fixed and (2) may have been related to the nginx configuration (I'm using Apache).
It is also identical to #898, but this issue is fixed long since, and the resolution suggested there (manually deleting the cookies in the affected browser on the client) does not work.

The only thing that helps is to restart the server -- meaning: the machine. Interestingly, restarting Apache alone (`sudo apachectl -k restart`) does not help.

## Logs
This incident creates 655 lines in the log. However, there are only 5 distinct lines:

Occurences | Line
---- | ---
327 lines | 2020-01-08 19:08:19   [ampache] (dba.class) -> Connection failed: SQLSTATE[HY000] [2002] No such   file or directory
292 lines |  2020-01-08 19:08:19 [ampache] (dba.class)   -> Error: failed to get database handle
1 lines |  2020-01-08 19:08:19 [ampache]   (access.class) -> check_network interface: ip not found in ACL   (client IP address)
1 lines |  2020-01-08 19:08:19 [ampache]   (login.class) -> UI::access_denied:(client IP address) is not in the Interface   Access list
34 lines |  2020-01-08 19:08:19 [ampache] (dba.class)   -> Wrong dbh.

Again, this is the same ACL error as in #2027, but I don't see how to leverage this here.
Any thoughts?

Thanks, M.

**Comment by AshotN** [comment]
It sounds to me like a server configuration issue. I would post your Nginx/Apache config file here.

**Comment by DrMMu** [comment]
Thank you for looking at this!

I have an Ubuntu server, so my Apache config is spread across multiple files (Debian style).

This is `apache2.conf` (starting with the first line which is not an introductory comment):
```
#
# The directory where shm and other runtime files will be stored.
#

DefaultRuntimeDir ${APACHE_RUN_DIR}

#
# PidFile: The file in which the server should record its process
# identification number when it starts.
# This needs to be set in /etc/apache2/envvars
#
PidFile ${APACHE_PID_FILE}

#
# Timeout: The number of seconds before receives and sends time out.
#
Timeout 300

#
# KeepAlive: Whether or not to allow persistent connections (more than
# one request per connection). Set to "Off" to deactivate.
#
KeepAlive On

#
# MaxKeepAliveRequests: The maximum number of requests to allow
# during a persistent connection. Set to 0 to allow an unlimited amount.
# We recommend you leave this number high, for maximum performance.
#
MaxKeepAliveRequests 100

#
# KeepAliveTimeout: Number of seconds to wait for the next request from the
# same client on the same connection.
#
KeepAliveTimeout 5


# These need to be set in /etc/apache2/envvars
User ${APACHE_RUN_USER}
Group ${APACHE_RUN_GROUP}

#
# HostnameLookups: Log the names of clients or just their IP addresses
# e.g., www.apache.org (on) or 204.62.129.132 (off).
# The default is off because it'd be overall better for the net if people
# had to knowingly turn this feature on, since enabling it means that
# each client request will result in AT LEAST one lookup request to the
# nameserver.
#
HostnameLookups Off

# ErrorLog: The location of the error log file.
# If you do not specify an ErrorLog directive within a <VirtualHost>
# container, error messages relating to that virtual host will be
# logged here.  If you *do* define an error logfile for a <VirtualHost>
# container, that host's errors will be logged there and not here.
#
ErrorLog ${APACHE_LOG_DIR}/error.log

#
# LogLevel: Control the severity of messages logged to the error_log.
# Available values: trace8, ..., trace1, debug, info, notice, warn,
# error, crit, alert, emerg.
# It is also possible to configure the log level for particular modules, e.g.
# "LogLevel info ssl:warn"
#
LogLevel warn

# Include module configuration:
IncludeOptional mods-enabled/*.load
IncludeOptional mods-enabled/*.conf

# Include list of ports to listen on
Include ports.conf


# Sets the default security model of the Apache2 HTTPD server. It does
# not allow access to the root filesystem outside of /usr/share and /var/www.
# The former is used by web applications packaged in Debian,
# the latter may be used for local directories served by the web server. If
# your system is serving content from a sub-directory in /srv you must allow
# access here, or in any related virtual host.
<Directory />
	Options FollowSymLinks
	AllowOverride None
	Require all denied
</Directory>

<Directory /usr/share>
	AllowOverride None
	Require all granted
</Directory>

<Directory /var/www/>
	Options Indexes FollowSymLinks
	AllowOverride All
	Require all granted
</Directory>

#<Directory /srv/>
#	Options Indexes FollowSymLinks
#	AllowOverride None
#	Require all granted
#</Directory>




# AccessFileName: The name of the file to look for in each directory
# for additional configuration directives.  See also the AllowOverride
# directive.
#
AccessFileName .htaccess

#
# The following lines prevent .htaccess and .htpasswd files from being
# viewed by Web clients.
#
<FilesMatch "^\.ht">
	Require all denied
</FilesMatch>


#
# The following directives define some format nicknames for use with
# a CustomLog directive.
#
# These deviate from the Common Log Format definitions in that they use %O
# (the actual bytes sent including headers) instead of %b (the size of the
# requested file), because the latter makes it impossible to detect partial
# requests.
#
# Note that the use of %{X-Forwarded-For}i instead of %h is not recommended.
# Use mod_remoteip instead.
#
LogFormat "%v:%p %h %l %u %t \"%r\" %>s %O \"%{Referer}i\" \"%{User-Agent}i\"" vhost_combined
LogFormat "%h %l %u %t \"%r\" %>s %O \"%{Referer}i\" \"%{User-Agent}i\"" combined
LogFormat "%h %l %u %t \"%r\" %>s %O" common
LogFormat "%{Referer}i -> %U" referer
LogFormat "%{User-agent}i" agent

# Include of directories ignores editors' and dpkg's backup files,
# see README.Debian for details.

# Include generic snippets of statements
IncludeOptional conf-enabled/*.conf

# Include the virtual host configurations:
IncludeOptional sites-enabled/*.conf

RedirectMatch ^/$ /ampache/

# vim: syntax=apache ts=4 sw=4 sts=4 sr noet
```

I think the two relevant files which are `included` from `sites-enabled` are the following, which deal with the config for port 80 (`http`) and 443 (`https`). The former only redirects to `https`:

```
<VirtualHost *:80>
	# The ServerName directive sets the request scheme, hostname and port that
	# the server uses to identify itself. This is used when creating
	# redirection URLs. In the context of virtual hosts, the ServerName
	# specifies what hostname must appear in the request's Host: header to
	# match this virtual host. For the default virtual host (this file) this
	# value is not decisive as it is used as a last resort host regardless.
	# However, you must set it for any further virtual host explicitly.
	#ServerName www.example.com

	ServerAdmin webmaster@localhost
	DocumentRoot /var/www/html

	# Available loglevels: trace8, ..., trace1, debug, info, notice, warn,
	# error, crit, alert, emerg.
	# It is also possible to configure the loglevel for particular
	# modules, e.g.
	#LogLevel info ssl:warn

	ErrorLog ${APACHE_LOG_DIR}/error.log
	CustomLog ${APACHE_LOG_DIR}/access.log combined

	# For most configuration files from conf-available/, which are
	# enabled or disabled at a global level, it is possible to
	# include a line for only one particular virtual host. For example the
	# following line enables the CGI configuration for this host only
	# after it has been globally disabled with "a2disconf".
	#Include conf-available/serve-cgi-bin.conf
RewriteEngine on
RewriteCond %{SERVER_NAME} = name.of.my.server.here
RewriteRule ^ https://%{SERVER_NAME}%{REQUEST_URI} [END,NE,R=permanent]
</VirtualHost>

# vim: syntax=apache ts=4 sw=4 sts=4 sr noet
```

The real deal is in this config for `https`:
```
<IfModule mod_ssl.c>
<VirtualHost *:443>
	# The ServerName directive sets the request scheme, hostname and port that
	# the server uses to identify itself. This is used when creating
	# redirection URLs. In the context of virtual hosts, the ServerName
	# specifies what hostname must appear in the request's Host: header to
	# match this virtual host. For the default virtual host (this file) this
	# value is not decisive as it is used as a last resort host regardless.
	# However, you must set it for any further virtual host explicitly.
	#ServerName www.example.com

	ServerAdmin webmaster@localhost
	DocumentRoot /var/www/html

	# Available loglevels: trace8, ..., trace1, debug, info, notice, warn,
	# error, crit, alert, emerg.
	# It is also possible to configure the loglevel for particular
	# modules, e.g.
	#LogLevel info ssl:warn

	ErrorLog ${APACHE_LOG_DIR}/error.log
	CustomLog ${APACHE_LOG_DIR}/access.log combined

	# For most configuration files from conf-available/, which are
	# enabled or disabled at a global level, it is possible to
	# include a line for only one particular virtual host. For example the
	# following line enables the CGI configuration for this host only
	# after it has been globally disabled with "a2disconf".
	#Include conf-available/serve-cgi-bin.conf


ServerName name.of.my.server.here
SSLCertificateFile /etc/letsencrypt/live/name.of.my.server.here/fullchain.pem
SSLCertificateKeyFile /etc/letsencrypt/live/name.of.my.server.here/privkey.pem
Include /etc/letsencrypt/options-ssl-apache.conf

#RedirectMatch ^/$ /ampache/
</VirtualHost>
</IfModule>
```

All other files which are `included` are standard, including `options-ssl-apache.conf` mentioned in the conf file above.

Does this inspire some ideas?

Thanks, M.

**Comment by AshotN** [comment]
I looked at the errors again and noticed that it’s having trouble connecting to your database. But you said the issue is sporadic? 


Check if your database is online first. 

@lachlan-00 will probably be more helpful whenever he has time to comment.

**Comment by lachlan-00** [comment]
I know why these are happening in 4 vs 3.

Ampache used to always make the same session id no matter what you did which is why this would never happen in 3. It also meant you could spoof a session from any of by getting that session id.

Your browser would have a cookie for session1234 and Ampache would always make the session id session1234. This meant that your user could log in anywhere and never get an access error.

Now that sessions are unique we have to look at how PHP and sessions are stored. It is something I think we can do in a better way or at least document better.

Short answer is that I think your session is cleaned by PHP and then your cookie is not kept on the server making the session invalid.

Long answer is we need to do some digging.

**Comment by DrMMu** [comment]
OK, thanks.
How can I be of help here?

As a workaround, would it help to always log off explicitly from ampache when I am using a browser?

Thanks, M.

**Comment by manuelglez86** [comment]
> 
> 
> I know why these are happening in 4 vs 3.
> 
> Ampache used to always make the same session id no matter what you did which is why this would never happen in 3. It also meant you could spoof a session from any of by getting that session id.
> 
> Your browser would have a cookie for session1234 and Ampache would always make the session id session1234. This meant that your user could log in anywhere and never get an access error.
> 
> Now that sessions are unique we have to look at how PHP and sessions are stored. It is something I think we can do in a better way or at least document better.
> 
> Short answer is that I think your session is cleaned by PHP and then your cookie is not kept on the server making the session invalid.
> 
> Long answer is we need to do some digging.

But when a session is cleaned Ampache just ask for your credentials again, it didn throws an error like that.

@DrMMu says that he restarts the Apache server but it looks like is an mysql problem, it seems like ampache can´t connect to the database for some reason, Yu have tried to reestart the sql server instead?.

**Comment by lachlan-00** [comment]
I see a possible issue in the api at least
![image](https://user-images.githubusercontent.com/1305249/73232820-4e3f5d80-41d0-11ea-9a85-4dabb6221c82.png)

**Comment by lachlan-00** [comment]
i'm going to do some more digging on this. as long as your browser has the cookie in it it should be kept by mysql during that time. which takes out php as an issue.

**Comment by manuelglez86** [comment]
maybe that is related to my case #2274

**Comment by DrMMu** [comment]
@manuelglez86 Thanks for suggesting to restart the `mysql` server instead!
# tl;dr
The server ran out of memory and terminated `mysqld`. I assume that this has nothing to do with ampache. I'll close this issue. Why this has never happened with ampache 3.9 is beyond me. For the benefit of others (possibly my future self), I describe what I did to diagnose (and hopefully solve) the problem. In the following.

# The long version
Since I could not reproduce the issue at will, I had to wait for it to reoccur, which it just did.
As suggested by @manuelglez86, I looked at the `mysql` server. And voila, `sudo /etc/init.d/mysql status -l` revealed that mysql was killed 1,5 days ago (I didn't notice that because I play my music using DSub in offline mode):

```
● mysql.service - MySQL Community Server
   Loaded: loaded (/lib/systemd/system/mysql.service; enabled; vendor preset: enabled)
   Active: failed (Result: exit-code) since Thu 2020-02-06 06:09:55 UTC; 1 day 11h ago
 Main PID: 970 (code=killed, signal=KILL)

Feb 06 06:09:55 ip-xxx-xxx-xxx-xxx systemd[1]: mysql.service: Failed with result 'exit-code'.
Feb 06 06:09:55 ip-xxx-xxx-xxx-xxx systemd[1]: Failed to start MySQL Community Server.
Feb 06 06:09:55 ip-xxx-xxx-xxx-xxx systemd[1]: mysql.service: Service hold-off time over, scheduling restart.
Feb 06 06:09:55 ip-xxx-xxx-xxx-xxx systemd[1]: mysql.service: Scheduled restart job, restart counter is at 10.
Feb 06 06:09:55 ip-xxx-xxx-xxx-xxx systemd[1]: Stopped MySQL Community Server.
Feb 06 06:09:55 ip-xxx-xxx-xxx-xxx systemd[1]: mysql.service: Start request repeated too quickly.
Feb 06 06:09:55 ip-xxx-xxx-xxx-xxx systemd[1]: mysql.service: Failed with result 'exit-code'.
Feb 06 06:09:55 ip-xxx-xxx-xxx-xxx systemd[1]: Failed to start MySQL Community Server.
```

mysql's log from 1.5 days was already gzipped away. It contained ten identical blocks (as mysql tried to restart 10 times, see the status report above) with the following errors at their core:
```
2020-02-06T06:09:55.028943Z 0 [ERROR] InnoDB: mmap(137428992 bytes) failed; errno 12
2020-02-06T06:09:55.028959Z 0 [ERROR] InnoDB: Cannot allocate memory for the buffer pool
2020-02-06T06:09:55.028966Z 0 [ERROR] InnoDB: Plugin initialization aborted with error Generic error
2020-02-06T06:09:55.028972Z 0 [ERROR] Plugin 'InnoDB' init function returned error.
2020-02-06T06:09:55.028977Z 0 [ERROR] Plugin 'InnoDB' registration as a STORAGE ENGINE failed.
2020-02-06T06:09:55.029005Z 0 [ERROR] Failed to initialize builtin plugins.
2020-02-06T06:09:55.029011Z 0 [ERROR] Aborting
```

So there wasn't enough memory for a plug in which ampache to the best of my knowledge does not use anyway so I googled how to get rid of it and found [this page](https://medium.com/@jm_c/mysql-crashing-innodb-mmap-bytes-failed-errno-12-64d5b801b2f2), where it is advised to check the syslog.

The syslog of 1.5 days ago was already gzipped away as well. The key lines are:
```
Feb  6 06:09:46 ip-xxx-xxx-xxx-xxx kernel: [434266.560377] Out of memory: Kill process 970 (mysqld) score 394 or sacrifice child
Feb  6 06:09:46 ip-xxx-xxx-xxx-xxx kernel: [434266.567760] Killed process 970 (mysqld) total-vm:1163124kB, anon-rss:193216kB, file-rss:0kB, shmem-rss:0kB
Feb  6 06:09:46 ip-xxx-xxx-xxx-xxx kernel: [434266.620047] oom_reaper: reaped process 970 (mysqld), now anon-rss:0kB, file-rss:0kB, shmem-rss:0kB
Feb  6 06:09:48 ip-xxx-xxx-xxx-xxx systemd[1]: mysql.service: Main process exited, code=killed, status=9/KILL
```

Then the syslog shows the 10 attempts to restart mysql, which fail.

As described in [the medium article](https://medium.com/@jm_c/mysql-crashing-innodb-mmap-bytes-failed-errno-12-64d5b801b2f2), I didn't have swap space (like in: not at all). I'm using an Amazon Lightsail server which [obviously comes without a swap partition out of the box](https://forums.aws.amazon.com/thread.jspa?threadID=251638), but up to version 3.9 of ampache I never seemed to need one, so this has never occurred to me.

I added a swap file as described [here](https://www.digitalocean.com/community/tutorials/how-to-add-swap-space-on-ubuntu-18-04) (linked in the medium article). I suppose this solves my problem, but I'll report if not.

Thanks everyone for helping me out!

Best, M.
